docs/glossary.md

```diff
@@ -0,0 +1,135 @@
+Frame types are created by `define-stack-frame` macro, which defines macros to push/pop specifically those global variables.
+
+Some general frame types (e.g. `ufun-frame`) serve the evaluator, and some serve specific primitives.
+For example `repeat` primitive keeps the remaining number of iterations in the global variable `*repeat-count*`, and that's one of the things the `repeat-special-frame` PDL frame saves/restores.
+
+### vpdl [from "Value Push Down List"]
+
+A stack of values which will become arguments to functions.
+
+The naive "copy and execute" model actually mutates the calling box, replacing calls with their return values, variables with their values etc.
+As an optimization when not stepping, this can mostly be skipped — and VPDL stack holds the intermediate values.
+
+It's also the interface to primitives.  `defboxer-primitive` looks similar to `defun` but constructs a lisp functions that pops arguments from VPDL, and pushes return value to VPDL.
+
+### recursive evaluation
+
+The evaluator calls primitives.  Primitives like `repeat` or `tell` that need to run Boxer code can't just call the evaluator (that would grow the Lisp stack and would be opaque to the stepper).
+
+`defrecursive-eval-primitive` / `defrecursive-funcall-primitive` macros split each such primitive into a before, after, and cleanup parts, plus a custom PDL frame type.
+Between the parts, control returns to the main eval loop, with state (e.g. `repeat`'s loop counter) kept in global variables, whose previous values are retained in the PDL frame.
+
+### process
+
+### sfun
+
+### ufun
+
+## vc - virtual copy
+
+A persistent data structure representing boxes in a way optimized for the evaluator.  It simulates deep copying & subsequent mutations while still sharing data.
+
+### multiple pointer
+
+The central idea in VC data structure is it gets forked _at the point of mutation_. A "single pointer" gets promoted to a "multiple pointer" with slots retaining both original and new value(s), each slot tagged by "who" (usually an ancestor of the box) and "when" made the change.
+
+`get-pointer-value` disambiguates the relevant values for a particalur copy & timestamp.
+
+### mps
+
+multiple pointer slot.
+
+## Variables in Boxer
+
+All this applies equally to variable and procedure lookup by name, the rules are the same.
+(Lookup finds a named box; only then you'll see if it's a data or doit box, and what to do with it.
+In this sense, Boxer falls into ["Lisp-1" camp](http://www.nhplace.com/kent/Papers/Technical-Issues.html).)
+
+To avoid confusion here, it's helpful to distinguish (1) the model users learn—and can observe by stepping (2) the semantics that arise from the model (3) the optimized implementation.
+
+### model: Lookup follows box structure
+
+To resolve a name, Boxer looks for a named box with same name (case insensitive).
+It first checks inside current box (including closet), and recursing into transparent boxes it contains.
```

Comment:
I wanted to say more about implementation of transparent boxes (I suspect the cost of sidewise-and-downward search was part of the motivation for adding "static variable cache"?) but wasn't certain, so for now only mentioned them here in the abstract model.